consensus/istanbul/core/roundchange_v2.go

```diff
@@ -0,0 +1,247 @@
+		if err != nil {
+			logger := c.newLogger("func", "sendRoundChange")
+			logger.Warn("Cannot build signed roundChangeV2 message", "error", err)
+			return
+		}
+		c.broadcast(msg)
+	} else {
+		c.broadcast(c.buildRoundChangeMsgV1(c.current.DesiredRound()))
+	}
+}
+
+// sendRoundChange sends a ROUND CHANGE message for the current desired round back to a single address
+func (c *core) sendRoundChangeAgain(addr common.Address) {
+	if c.isConsensusFork(c.current.Sequence()) {
+		msg, err := c.buildSignedRoundChangeMsgV2(c.current.DesiredRound())
+		if err != nil {
+			logger := c.newLogger("func", "sendRoundChangeAgain", "addr", addr)
+			logger.Warn("Cannot build signed roundChangeV2 message", "error", err)
+			return
+		}
+		c.unicast(msg, addr)
+	} else {
+		c.unicast(c.buildRoundChangeMsgV1(c.current.DesiredRound()), addr)
+	}
+}
+
+// buildRoundChangeV2 builds a roundChangeV2 instance with an empty prepared certificate
+func buildRoundChangeV2(addr common.Address, view *istanbul.View) *istanbul.RoundChangeV2 {
+	pc, proposal := istanbul.EmptyPreparedCertificateV2()
+	return &istanbul.RoundChangeV2{
+		Request: istanbul.RoundChangeRequest{
+			Address:               addr,
+			View:                  *view,
+			PreparedCertificateV2: pc,
+		},
+		PreparedProposal: proposal,
+	}
+}
+
+// buildSignedRoundChangeMsgV2 builds a roundChangeV2 istanbul.Message, with the inner
+// roundChangeRequest properly signed, ready to be broadcast/unicast.
+func (c *core) buildSignedRoundChangeMsgV2(round *big.Int) (*istanbul.Message, error) {
+	nextView := &istanbul.View{
+		Round:    new(big.Int).Set(round),
+		Sequence: new(big.Int).Set(c.current.View().Sequence),
+	}
+	roundChangeV2 := buildRoundChangeV2(c.address, nextView)
+	pc := c.current.PreparedCertificate()
+	if !pc.IsEmpty() {
+		// Add prepare certificate proposal and votes
+		roundChangeV2.Request.PreparedCertificateV2 = istanbul.PCV2FromPCV1(pc)
+		roundChangeV2.PreparedProposal = pc.Proposal
+	}
+	// Sign the round change request
+	if err := roundChangeV2.Request.Sign(c.backend.Sign); err != nil {
+		return nil, err
+	}
+	return istanbul.NewRoundChangeV2Message(roundChangeV2, c.address), nil
+}
+
+func (c *core) handleRoundChangeCertificateV2(view istanbul.View, roundChangeCertificateV2 istanbul.RoundChangeCertificateV2, proposal istanbul.Proposal) error {
+	subject := istanbul.Subject{
+		View:   &view,
+		Digest: proposal.Hash(),
+	}
+
+	logger := c.newLogger("func", "handleRoundChangeCertificateV2", "proposal_round", subject.View.Round, "proposal_seq", subject.View.Sequence, "proposal_digest", subject.Digest.String())
+
+	if len(roundChangeCertificateV2.Requests) > c.current.ValidatorSet().Size() || len(roundChangeCertificateV2.Requests) < c.current.ValidatorSet().MinQuorumSize() {
+		return errInvalidRoundChangeCertificateNumMsgs
+	}
+
+	seen := make(map[common.Address]bool)
+	for i := range roundChangeCertificateV2.Requests {
```

Comment:
This is something that I don't have a proper answer for. I remember checking the paper for specifically this question, and we discussed this with @gastonponti , about if it makes sense or not to do it. In theory only the prepared certificate for the highest round should be used to match the proposal. So that's the only one that I'm verifying (below, at `preparedView, err := c.verifyPCV2WithProposal(*pc, proposal)` ).

Reply:
Having thought more about this, i think this is safe.

A malicious proposer could include f+1 round change requests that are not valid (fake proposals, no prepare or commit messages), but they cannot forge a valid round change request with prepared certificate because 2f+1 prepare or commit messages would be needed for that.

So if a malicious proposer included a fake round change request with prepared certificate it will not pass verification, nodes will not vote for the proposal and the effect would be as if the proposer had not proposed.

Reply:
The proposer cannot actually do that, because we are checking the signatures of all the requests. So what is not being checked here is if the prepared certificate signed by the emitter of the roundChangeRequest is valid. But we know that the request is properly signed